src/globus_sdk/client.py

```diff
@@ -91,16 +102,86 @@ def __init__(
         self.transport = self.transport_class(**(transport_params or {}))
         log.debug(f"initialized transport of type {type(self.transport)}")
 
+        if app and (authorizer or app_name):
+            raise exc.GlobusSDKUsageError(
+                f"A {type(self).__name__} cannot accept both an 'app' and an 'app_name'"
+                " or 'authorizer' as these values come from the 'app' instead."
+            )
+
+        self._app = app
         self.authorizer = authorizer
 
-        # set application name if given
+        # set application name if available from app or app_name
         self._app_name = None
         if app_name is not None:
             self.app_name = app_name
+        if app is not None:
+            self.app_name = app.app_name
 
         # setup paginated methods
         self.paginated = PaginatorTable(self)
 
+        self._finalize_app()
+
+    @property
+    def default_scope_requirements(self) -> list[Scope]:
+        """
+        Scopes that will automatically be added to this client's app's
+        scope_requirements during _finalize_app.
+
+        For clients with static scope requirements this can just be a static
+        value. Clients with dynamic requirements should use @property and must
+        return sane results while the Base Client is being initialized.
+        """
+        raise NotImplementedError
+
+    def _finalize_app(self) -> None:
+        if self._app:
+            if self.resource_server is None:
+                raise ValueError(
+                    "Unable to use an 'app' with a client with no "
+                    "'resource_server' defined."
+                )
+
+            self._app.add_scope_requirements(
+                {self.resource_server: self.default_scope_requirements}
+            )
+
+    def add_app_scope(self, scope_collection: ScopeCollectionType) -> None:
```

Comment:
Since AppRegistrationParams were omitted, could this scope adding function return the client instead of None for chained function calls?

I'd recommend we rename to  to `with_app_scope(...)` to indicate this but thinking about usage like.

```
flows = FlowsClient(app=app)\
   .with_app_scope(FlowsScopes.manage_flows)\
   .with_app_scope(FlowsScopes.run_manage)
```

Reply:
Definitely like chaining it! I'll plan to do the same to `TransferClient.add_app_data_access_scope`
Think I prefer keeping the name as is though.

As a chain `add_app_scope` still reads fine to me:
```
flows = FlowsClient(app=app)\
   .add_app_scope(FlowsScopes.manage_flows)\
   .add_app_scope(FlowsScopes.run_manage)
```

but when not chained, `with_app_scope` seems kinda strange to me:
```
try:
    client.do_the_thing()
except ConsentRequired as consent_error:
    client.with_app_scope(consent_error.required_scope)
```